src/classes/CRLP_DefaultConfigBuilder_SVC.cls

```diff
@@ -802,16 +808,19 @@ public class CRLP_DefaultConfigBuilder_SVC {
                 rollup.intValue = Integer.valueOf(rollupDetails[6]);
             }
 
+            String membershipRecordTypeDeveloperNames = buildMembershipRecordTypeDeveloperNames();
             if (CRLP_DefaultConfigBuilder.legacySettings.npo02__Enable_Soft_Credit_Rollups__c == false &&
-                    rollupDetails[2] == CRLP_DefaultConfigBuilder.pscAmountFld) {
+                rollupDetails[2] == CRLP_DefaultConfigBuilder.pscAmountFld) {
                 // If the legacy 'Enable Soft Credit Rollups' field is false,
                 // set all the soft credit rollup definitions to inactive.
                 rollup.isActive = false;
-            } else if (CRLP_DefaultConfigBuilder.legacySettings.npo02__Membership_Record_Types__c == null &&
-                    rollupDetails[1] == CRLP_DefaultConfigBuilder.FilterGroup_Membership) {
-                // If the legacy Membership Record Type setting is null,
-                // set all the membership rollup definitions to inactive.
-                rollup.isActive = false;
+
+            } else if (rollupDetails[1] == CRLP_DefaultConfigBuilder.FilterGroup_Membership
+                && String.isEmpty(membershipRecordTypeDeveloperNames)) {
+                // If the legacy Membership Record Type setting does not match existing record types,
+                // skip building rollup
```

Comment:
Why not create the rollup definition as inactive rather than not create at all?

Reply:
When creating the Filter Group `Memberships: Won`, I put a check in to see if there are membership record types related to what's in `npo02__Households_Settings__c.npo02__Membership_Record_Types__c`. If there are no existing record types, then I don't create the Filter Group at all, so here I have to make sure not to create the Rollups that lookup to the Filter Group or an error will get thrown. (See [ln 342](https://github.com/SalesforceFoundation/Cumulus/pull/3846/files/f4f37413ec8afe91f7dded3c430cfdce26bbf341#diff-9f2e945687d30091f67cb34ab2557e90R342) for where change is made).

Alternatively we can always create the Filter Group `Memberships: Won` and just not create the Filter Rule associated to the record types when the record types doesn't exist. Originally this is how it was set up:

Filter Group: `Memberships: Won`
Filter Rule: `Membership: Won Only`
Filter Rule: `Membership: Amount Not Null`
Filter Rule: `Membership: for Record Types` -> Only created if `CRLP_DefaultConfigBuilder.legacySettings.npo02__Membership_Record_Types__c != null`

If I do that instead, then the rollups can be created and set to Inactive. It just seemed weird to create a filter group that referred to Membership when there were no membership record types, but maybe that is the intended behavior

Reply:
Thanks for the explanation @seanpat09. I see why it would make sense to not create them, but if they're missing it leaves admins in an odd situation if they ever want to enable Membership rollups of some kind. I think we should do the latter - create the Filter Group with the IsWon/NotNull rules along with the associated (inactive) rollups. That allows an Admin to easily implement/activate membership rollups at any point in the future.

Reply:
@seanpat09 I want to be sure I am taking everything into account when testing this PR.  How does Membership: Won Filter Group come into play here?  I deleted the Membership record type from the Opportunity object.  Should there still be a Membership: Won Filter Group after that is done?  The Membership: Won Filter Group that I had in my scratch org did not have a Filter Rule = Opportunity - Record Type ID - In List - null .  Did not exist.  Did I need to add that first before deleting the Membership record type and enabling CRLP's?  When I add that Filter Rule to the Memberships: Won Filter Group, the Value field is required; I can't leave it blank.  Is this a result of the changes you made?  Do you know how it was not being populated previously?
![w-029366_crlp_filter groups_membership](https://user-images.githubusercontent.com/40002399/49967061-cc1ea780-feef-11e8-9a53-23f41b26ee3d.png)

Reply:
@rponti-sforg 
1) The filter group `Membership: Won` should always be created, regardless of whether or not the  Membership record type exists.
2) If after creating a scratch org you delete the Membership profile, but before enabling CRLP, then the Filter Rule that references Record Type ID should not be created at all. Before what was happening is that the Filter Rule would get created, but the value would be Null because the record type doesn't exist

Reply:
> @rponti-sforg
> 
> 1. The filter group `Membership: Won` should always be created, regardless of whether or not the  Membership record type exists.
> 2. If after creating a scratch org you delete the Membership profile, but before enabling CRLP, then the Filter Rule that references Record Type ID should not be created at all. Before what was happening is that the Filter Rule would get created, but the value would be Null because the record type doesn't exist

@seanpat09 Thank you Sean.  These are my observations:

1. Create scratch org;
2. Delete Membership record type from Opportunity object;
3. Check Membership: Won Filter Group.  Only two Opportunity rules listed; one for Won and one for Amount;
4. Enable CRLPs;
5. Create Opportunity off Contact;
6. Check Error Log.  No Errors observed;
7.  Check Membership: Won Filter Group again.  Same two rules listed as stated in #3 above.  No additional Opportunity rules created with value = null.

Is this the main test scenario and result we are trying to verify?

Reply:
@rponti-sforg yes, that is correct. @force2b are there any other scenarios you think we should look out for?

Reply:
Thank you both for the thorough work and testing on this. There are some additional items I'd consider:

- Don't delete any record types until after going into NPSP Settings and updating the legacy rollup record type exclusions so that one or more of them exclude the Membership Record Type. The idea is to have that record type Id referenced by the settings before deleting the record types and enabling CRLP.
- This issue is not unique to the Membership Record Type. I suggest excluding and deleting a different record type, such as "In Kind" (as an additional test) to confirm that the behavior is still as expected.
- There are 3 types of record type exclusions in NPSP Settings: Account, Contact, and GAU. Any or all of these 3 could be updated in NPSP Settings to exclude the record type(s) to be deleted to test that the rule creation logic used during CRLP Enablement ignores the record type id if it doesn't exist.

Reply:
> Thank you both for the thorough work and testing on this. There are some additional items I'd consider:
> 
> * Don't delete any record types until after going into NPSP Settings and updating the legacy rollup record type exclusions so that one or more of them exclude the Membership Record Type. The idea is to have that record type Id referenced by the settings before deleting the record types and enabling CRLP.

**Thank you Mike.  Re-testing in new scratch org based on your feedback above.**  

> * This issue is not unique to the Membership Record Type. I suggest excluding and deleting a different record type, such as "In Kind" (as an additional test) to confirm that the behavior is still as expected.

**Perfect.  Testing this scenario as well in new scratch org.**

> * There are 3 types of record type exclusions in NPSP Settings: Account, Contact, and GAU. Any or all of these 3 could be updated in NPSP Settings to exclude the record type(s) to be deleted to test that the rule creation logic used during CRLP Enablement ignores the record type id if it doesn't exist.

**This last scenario, can that be tested in PR#3844 ...after PR#3846 is merged into PR#3844?  Is the last scenario the same as what we were testing with PR#3844?**

Reply:
@rponti-sforg The last scenario is the same as what we were testing in #3844, but just with other use cases. Since this is a child branch of #3844, everything in #3844 is in #3846, so you should be able to do those tests in this branch as well.

Reply:
> > Thank you both for the thorough work and testing on this. There are some additional items I'd consider:
> > 
> > * Don't delete any record types until after going into NPSP Settings and updating the legacy rollup record type exclusions so that one or more of them exclude the Membership Record Type. The idea is to have that record type Id referenced by the settings before deleting the record types and enabling CRLP.
> 
> **Thank you Mike. Re-testing in new scratch org based on your feedback above.**
> 
> > * This issue is not unique to the Membership Record Type. I suggest excluding and deleting a different record type, such as "In Kind" (as an additional test) to confirm that the behavior is still as expected.
> 
> **Perfect. Testing this scenario as well in new scratch org.**
> 
> > * There are 3 types of record type exclusions in NPSP Settings: Account, Contact, and GAU. Any or all of these 3 could be updated in NPSP Settings to exclude the record type(s) to be deleted to test that the rule creation logic used during CRLP Enablement ignores the record type id if it doesn't exist.
> 
> **This last scenario, can that be tested in PR#3844 ...after PR#3846 is merged into PR#3844? Is the last scenario the same as what we were testing with PR#3844?**

@force2b @seanpat09 While my testing did not produce any **null pointer exception errors** or **Membership filter rules** containing **Record Type ID** with **Value = null**, I wanted to confirm exactly what we should be seeing within **CLRP > Filter Groups**.  Specifically, rules that contain **Record Type ID** and a **Value = null**.  

My first question is, are we just concerned with the **Memberships: Won** Filter Group?
![default filter groups](https://user-images.githubusercontent.com/40002399/50130567-32872b00-024c-11e9-932e-6f63176bed6d.png)
  
Second, for the **Opps: Won (HC) Excl RecTypes** Filter Group, I am seeing a rule that includes **Record Type ID** and **Value = null**.  Is this something we need to be concerned about?  I also see this in the **Opps: Won (SC) Excl RecTypes, Pmts: Paid Excl Rec Types, Pmts: WrittenOff Excl Rec Types** .  

Third, it looks like a User can exclude opp record types for **Payments**.  I tested Payments the same way I tested Accounts, Contacts, and GAUs record type exclusions (minus the Recalculate Rollups function).  Do we need to be concerned with this object and the exclusion of record types on it?
![payments_exclude opp record types](https://user-images.githubusercontent.com/40002399/50130591-521e5380-024c-11e9-8393-4ea14075f8b3.png)

Reply:
@rponti-sforg Thanks for testing and for the great details/questions

1. I confirm that we should **not** be seeing a Filter Rule with the following. What should happen is that this rule (just this one rule within the group) shouldn't exist. As with before this isn't specific to the Membership Record Type. A rule like the one below shouldn't be created at all, regardless of the name of the Rule or the original record type it was intended to reference.
> Record Type ID _[Not in List]_ null
2. Good point on Payments. I'm pretty sure those follow the same logic as Opportunities when creating the Filter Groups & Rules, but it's worth validating.

There's a lot of code in the method that creates these, and it's grown since originally written so I can definitely see where this is a game of whack-a-mole. The good thing is that the parent branch for this PR fixes the bug that caused an exception when attempting to evaluate this rule. We're just trying to avoid creating the offending rule in the first place. The overall intent of this PR is that rule(s) such as the one below will never be created for any Filter Group relating to any Rollup or Record Type.
> Record Type ID _[Not in List]_ null
> Record Type ID _[In List]_ null

Just a point of clarification, technically the following rule is perfectly valid:
> Record Type ID _[Equals]_ null

Reply:
> @rponti-sforg Thanks for testing and for the great details/questions
> 
> 1. I confirm that we should **not** be seeing a Filter Rule with the following. What should happen is that this rule (just this one rule within the group) shouldn't exist. As with before this isn't specific to the Membership Record Type. A rule like the one below shouldn't be created at all, regardless of the name of the Rule or the original record type it was intended to reference.
> 
> > Record Type ID _[Not in List]_ null
> 
> 1. Good point on Payments. I'm pretty sure those follow the same logic as Opportunities when creating the Filter Groups & Rules, but it's worth validating.
> 
> There's a lot of code in the method that creates these, and it's grown since originally written so I can definitely see where this is a game of whack-a-mole. The good thing is that the parent branch for this PR fixes the bug that caused an exception when attempting to evaluate this rule. We're just trying to avoid creating the offending rule in the first place. The overall intent of this PR is that rule(s) such as the one below will never be created for any Filter Group relating to any Rollup or Record Type.
> 
> > Record Type ID _[Not in List]_ null
> > Record Type ID _[In List]_ null
> 
> Just a point of clarification, technically the following rule is perfectly valid:
> 
> > Record Type ID _[Equals]_ null

@force2b @seanpat09 Thanks Mike for the input.  Helps me to put the pieces together and see the bigger picture.  How do we want to proceed?

1. My testing did not expose or surface any null pointer exception errors for when specific record types are excluded from Opp record types and subsequently deleted from the Opp object itself;

2. Record Type ID filter rules where Value = null are still being created within other Filter Groups.  As you state, this should not occur regardless of type of Filter Group.  However, it is not being created in Membership: Won Filter Group (this was the original issue reported).  Do we keep PR3844 and 46 open and the subsequent WI to address the Record Type ID filter rules still being created with Value = null?

Reply:
@rponti-sforg Let me look into the other record type filter groups where value is still being created as null, I thought I had accounted for this in all the filter group types